coverage: Switch to raw coverage files

anomiex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
It turns out that there really aren't any really good generic coverage reporting tools, the best you can get is something that highlights covered lines without function/method stats, branch info, or anything else fancy.

To provide useful coverage output, we're going to look instead at generating separate coverage dashboards for PHP (PHPUnit) and JS (jest/istanbul), each generated by those tools so we can get all the useful info possible. To do that, we'll need all the coverage commands to generate the tools' raw data formats rather than a lossy format like clover.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

*

anomiex committed:
coverage: Switch to raw coverage files

It turns out that there really aren't any really good generic coverage
reporting tools, the best you can get is something that highlights
covered lines without function/method stats, branch info, or anything
else fancy.

To provide useful coverage output, we're going to look instead at
generating separate coverage dashboards for PHP (PHPUnit) and JS
(jest/istanbul), each generated by those tools so we can get all the
useful info possible. To do that, we'll need all the coverage commands
to generate the tools' raw data formats rather than a lossy format like
clover.

anomiex committed:
lockfiles

anomiex committed:
Debugging

tbradsha commented on docs/monorepo.md:
```diff
@@ -309,7 +309,9 @@ If a project contains PHP or JavaScript tests, it should also define `.scripts.t
-For PHP tests, you'll probably run PHPUnit as `php -dpcov.directory=. "$(command -v phpunit)" --coverage-clover "$COVERAGE_DIR/php/clover.xml"`.
+For PHP tests, you'll probably run PHPUnit as `php -dpcov.directory=. "$(command -v phpunit)" --coverage-php "$COVERAGE_DIR/php.cov"`. If you have multiple runs (e.g. unit and integration), be sure to write the `php.cov` files to separate subdirectories of `$COVERAGE_DIR`.
+
+For JS tests, you'll probably have a `test` script in package.json that runs `jest` with any needed options, and then a `test-coverage` script that does `pnpm run test --coverage --coverageDirectory=\"$COVERAGE_DIR\" --coverageReporters=json`. If you have multiple runs (e.g. unit and integration), be sure each run writes to a different subdirectory of `$COVERAGE_DIR`.
```
For `jest`, should we explicitly name the output file like we do PHP? What is the default filename?

anomiex replied:
To name the file you'd have to do like this in a jest.config.js, I think
```
coverageReporters: [
    [ 'json', { file: 'whatever.json' } ]
]
```
which I guess we could reasonably do now that I had to add a jest.config.js everywhere.

The default name is `coverage-final.json`.

tbradsha commented on .github/workflows/tests.yml:
```diff
@@ -218,6 +218,10 @@ jobs:
                   FAIL=true
                 fi
 
+                # XXX
+                find "$COVERAGE_DIR" -type f -exec ls -lh {} +
+                df -h .
+
```
Obviously this is just debug, but noting it's giving errors on the non-coverage tests since `$COVERAGE_DIR` isn't set.

anomiex replied:
Removed now. I had wondered if it was blowing up because the files were too big, but it turned out that wasn't it. It was OOM-ing because of some bad `collectCoverageFrom` in a few packages trying to process `wordpress/` and so on.

anomiex committed:
Set collectCoverageFrom correctly everywhere

This requires that we add jest.config.cjs to a bunch of projects that
didn't already have one.

By default the base config will include non-test files under `src/`.
Various packages need additional dirs such as `components/`, and a
root-level `index.jsx` or `global.d.ts` seems common too.

anomiex committed:
remove debug

tbradsha commented on projects/plugins/jetpack/tests/action-test-coverage.sh:
```diff
@@ -13,11 +13,11 @@ done
+TESTS[php-lfs]="WP_TESTS_CONFIG_FILE_PATH=\"${WP_TESTS_CONFIG_FILE_PATH%.php}.lfs.php\" LEGACY_FULL_SYNC=1 php -dpcov.directory=. \"$(command -v phpunit)\" --group=legacy-full-sync --coverage-php \"$COVERAGE_DIR/legacy-sync/php.cov\""
+TESTS[php-multisite]="WP_TESTS_CONFIG_FILE_PATH=\"${WP_TESTS_CONFIG_FILE_PATH%.php}.multi.php\" WP_MULTISITE=1 php -dpcov.directory=. \"$(command -v phpunit)\" -c tests/php.multisite.xml --coverage-php \"$COVERAGE_DIR/multisite/php.cov\""
+TESTS[client]="pnpm run test-client --coverage --coverageDirectory=\"$COVERAGE_DIR/client\" --coverageReporters=json"
+TESTS[gui]="pnpm run test-gui --coverage --coverageDirectory=\"$COVERAGE_DIR/client\" --coverageReporters=json"
+TESTS[extensions]="pnpm run test-extensions --coverage --coverageDirectory=\"$COVERAGE_DIR/client\" --coverageReporters=json"
```
```suggestion
TESTS[gui]="pnpm run test-gui --coverage --coverageDirectory=\"$COVERAGE_DIR/gui\" --coverageReporters=json"
TESTS[extensions]="pnpm run test-extensions --coverage --coverageDirectory=\"$COVERAGE_DIR/extensions\" --coverageReporters=json"
```

tbradsha replied:
Looks like we got the wrong dir names here.

anomiex replied:
Good catch

tbradsha reviewed (commented):
Maybe we can move the extra `jest` flags to the `jest` config, given we have the config files in most/all relevant projects.

Also made a comment inline regarding a typo.

anomiex committed:
Move `coverageDirectory` and `coverageReporters` into configs, now that we use configs everywhere.

anomiex committed:
Fix autoloader munger to not output clover

tbradsha reviewed (approved):
Looks good!
